compiler-plugin/src/main/scala/scalapb/compiler/ProtobufGenerator.scala

```diff
@@ -709,20 +712,28 @@ class ProtobufGenerator(
           else if (field.isRepeated) Some(s"${field.emptyCollection}")
           else None
 
-        ConstructorField(
-          name = field.scalaName.asSymbol,
-          typeName = typeName,
-          default = ctorDefaultValue,
-          annotations = annotations(field)
+        Seq(
+          ConstructorField(
+            name = field.scalaName.asSymbol,
+            typeName = typeName,
+            default = ctorDefaultValue,
+            annotations = annotations(field)
+          )
         )
+
+      case oneOfField if oneOfField.isInOneof =>
+        val oneOf = oneOfField.getContainingOneof
+        if (!visitedOneOffsNames.contains(oneOf.getFullName)) {
```

Comment:
instead of this approach, maybe we can use `sortBy` to sort the list of `ConstructorField`s  (`regularFeilds ++ oneofFields`). We can have two different key functions and it will be picked based on the opt-out value. We probably need to add to `ConstructorField` the necessary info for the sorting functions (like field index, for one ofs it can be the index of first field)